pkg/cake/image_tree.go

```diff
@@ -159,8 +160,36 @@ func WalkBuildGraph(graph *Image, apply func(image *Image)) {
 		if len(queue) == 0 {
 			break
 		}
-		image := queue[len(queue)-1]
+		image := queue[0]
+		queue[0] = nil
+		queue = queue[1:]
 		apply(image)
-		queue = append(image.Children, queue[:len(queue)-1]...)
+		queue = append(queue, image.Children...)
+	}
+}
```

Comment:
just wondering: do we still need to have this method or can we use `WalkBuildGraphParallel` instead?

Reply:
As both methods take a function as a parameter, it is the responsibility of a caller to implement proper synchronization. If the function passed to `WalkBuildGraphParallel` aggregates some data in some form of collections, it causes race conditions and inconsistent results. An example of this is report generation in `GenerateReport`. I'd prefer having a sequential version which preserves the order of traversal and parallel version which is used when deterministic order is not important.